src/Database/Esqueleto/Internal/Internal.hs

```diff
@@ -1797,15 +1872,49 @@ instance Monoid DistinctClause where
 data FromClause =
     FromStart Ident EntityDef
   | FromJoin FromClause JoinKind FromClause (Maybe (SqlExpr (Value Bool)))
+  | FromJoinPartial JoinKind FromClause
+  | FromJoinPartials [FromClause]
   | OnClause (SqlExpr (Value Bool))
   | FromQuery Ident (IdentInfo -> (TLB.Builder, [PersistValue]))
+  | FromMany [FromClause]
+  | FromNone
+
+
+-- Make the semigroup version of FromClause look like the old list version
+getFromClauseList :: FromClause -> [FromClause]
+getFromClauseList (FromMany fs) = fs
+getFromClauseList FromNone        = []
+getFromClauseList f             = [f]
+
+instance Semigroup FromClause where
+  FromNone <> f = f
+  f <> FromNone = f
+  lhs@(FromJoin _ _ _ _)  <> FromJoinPartial joinKind fromClause    = FromJoin lhs joinKind fromClause Nothing
+  lhs@(FromStart _ _)     <> FromJoinPartial joinKind fromClause    = FromJoin lhs joinKind fromClause Nothing
+  lhs@(FromQuery _ _)     <> FromJoinPartial joinKind fromClause    = FromJoin lhs joinKind fromClause Nothing
+  p@(FromJoinPartial _ _) <> p'@(FromJoinPartial _ _)               = FromJoinPartials [p, p']
+  p@(FromJoinPartial _ _) <> FromJoinPartials partials              = FromJoinPartials (p:partials)
+  f                       <> FromJoinPartials partials              = foldl (<>) f partials
+  f                       <> FromMany (f'@(FromJoinPartials _ ):fs) = FromMany ((f <> f') : fs)
+  f                       <> FromMany (f'@(FromJoinPartial _ _):fs) = FromMany ((f <> f') : fs)
+  FromMany fs             <> FromMany gs                            = FromMany (fs <> gs)
+  f                       <> FromMany fs                            = FromMany (f:fs)
+  FromMany fs             <> f                                      = FromMany (fs ++ [f])
+  f                       <> f'                                     = FromMany [f, f']
+
+instance Monoid FromClause where
+  mempty = FromNone
+  mappend = (<>)
```

Comment:
Have you verified that this is lawful? I think the first two clauses in the `Semigroup` instance should cover it, but I'd really prefer to see a machine-checked verification (eg QuickCheck or similar).

Reply:
No clue how to verify if a monoid is lawful